<a href="https://colab.research.google.com/github/Harsh9304/Machine-Learning/blob/main/Complete_NUMPY___PANDAS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

NUMPY - NUMERICAL PYTHON

Advantages of Numpy Arrys

1.Allows several Mathematical operations

2.Faster Operations

In [ ]:
import numpy as np


LIST VS NUMPY -tIME TAKEN

In [ ]:
from time import process_time

Time taken by list

In [ ]:
python_list =[i for i in range(1000000)]

stat_time=process_time()

python_list=[i+5 for i in python_list]

end_time=process_time()

print(end_time-stat_time)

0.06760363699999994


In [ ]:
np_array=np.array([i for i in range(1000000)])

start_time=process_time()

np_array+=5

end_time=process_time()

print(end_time-stat_time)


0.1868490679999999


SO USING A NUMPY ARRAY VS LIST,TUPLE,SET IS THAT IN ML WE WORK WITH LARGE NUMBER OF DATASET HAVING MILLION DATA POINTS SO IN THAT CASE OUR OPERTATIONS ARE FASTER AS IT TAKES LESS TIME BUT LIST TUPLE TAKES MORE TIME FOR THAT

NUMPY ARRAYS


In [ ]:
# LIST
list_1=[1,2,3,4,5]
print(list_1)
type(list_1)

[1, 2, 3, 4, 5]


list

In [ ]:
np_array=np.array([1,2,3,4,5])
print(np_array)   # so it doesn't contains comma it is like matrix in maths
type(np_array)

[1 2 3 4 5]


numpy.ndarray

In [ ]:
#creating a 1 D array
a=np.array([1,2,3,4])
print(a)

[1 2 3 4]


In [ ]:
a.shape #contains 4 columns

(4,)

In [ ]:
b=np.array([(1,2,3,4),(5,6,7,8)])
print(b) #2d array of 2 rows and 4 columns

[[1 2 3 4]
 [5 6 7 8]]


In [ ]:
b.shape

(2, 4)

In [ ]:
c=np.array([(1,2,3,4),(5,9,8,7)],dtype=float)
print(c)

[[1. 2. 3. 4.]
 [5. 9. 8. 7.]]


Initial Placeholders in numpy arrays


In [ ]:
#create a numpy arrays of zeros
x=np.zeros((4,5))
print(x)

[[0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]]


In [ ]:
#create a numpy array of ones
y=np.ones((3,3))
print(y)

[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


In [ ]:
#array of particular value
z=np.full((5,4),5)
print(z)

[[5 5 5 5]
 [5 5 5 5]
 [5 5 5 5]
 [5 5 5 5]
 [5 5 5 5]]


In [ ]:
#create a identity matrix # it should be a square matrix i.e n*n
a=np.eye(5)
print(a)

[[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]]


In [ ]:
#create a numpy array with random values
b=np.random.random((3,4))
print(b)

[[0.67405593 0.22540604 0.829935   0.96974716]
 [0.45841052 0.5081118  0.89555773 0.17545745]
 [0.26114844 0.86860775 0.09092341 0.65647694]]


In [ ]:
#random integer value array
c=np.random.randint(10,100,(3,5))
print(c)

[[16 11 99 22 51]
 [34 42 24 74 52]
 [37 41 74 14 72]]


In [ ]:
#array of evenly spaced values---> specifying the number of values required
d=np.linspace(10,30,6)
print(d)

[10. 14. 18. 22. 26. 30.]


In [ ]:
#array of evenly spaced values---> specifying the step
e=np.arange(10,30,5)
print(e)

[10 15 20 25]


In [ ]:
#convert a list to a numpry array   -----> we can convert tuple set into numpy array also
list2=[10,20,30,40,50]

np_array=np.asarray(list2)
print(np_array)
type(np_array)

[10 20 30 40 50]


numpy.ndarray

Analysing a numpy array

In [ ]:
c=np.random.randint(10,90,(5,5))
print(c)

[[71 25 53 70 42]
 [12 63 42 56 29]
 [36 70 83 81 44]
 [46 38 76 35 27]
 [14 68 63 56 38]]


In [ ]:
#array dimension
print(c.shape)

(5, 5)


In [ ]:
#number of dimensions
print(c.ndim) #2 d array

2


In [ ]:
#number of elements in an array
print(c.size)

25


In [ ]:
#checking the data type of the values in the array
print(c.dtype)

int64


MATHEMATICAL OPERATIONS ON A NP ARRAY

In [ ]:
list_1=[1,2,3,4,5]
list_2=[6,7,8,9,10]
print(list_1+list_2) #in case of list it will only concatenate not add but numpy array do all mathematical operations

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [ ]:
a=np.random.randint(0,10,(3,3))
b=np.random.randint(10,20,(3,3))
print(a)
print(b)

[[6 5 1]
 [8 4 9]
 [8 5 6]]
[[18 19 10]
 [14 10 13]
 [17 16 12]]


In [ ]:
print(a+b)
print(a-b)
print(a*b)
print(a/b)

[[24 24 11]
 [22 14 22]
 [25 21 18]]
[[-12 -14  -9]
 [ -6  -6  -4]
 [ -9 -11  -6]]
[[108  95  10]
 [112  40 117]
 [136  80  72]]
[[0.33333333 0.26315789 0.1       ]
 [0.57142857 0.4        0.69230769]
 [0.47058824 0.3125     0.5       ]]


In [ ]:
a=np.random.randint(0,10,(3,3))
b=np.random.randint(10,20,(3,3))
print(a)
print(b)

[[3 4 4]
 [5 4 4]
 [6 4 0]]
[[18 17 10]
 [13 15 10]
 [14 15 16]]


In [ ]:
print(np.add(a,b))

[[21 21 14]
 [18 19 14]
 [20 19 16]]


In [ ]:
print(np.subtract(b,a))
print(np.subtract(a,b))
print(np.multiply(a,b))
print(np.divide(a,b))
print(np.divide(b,a))

[[15 13  6]
 [ 8 11  6]
 [ 8 11 16]]
[[-15 -13  -6]
 [ -8 -11  -6]
 [ -8 -11 -16]]
[[54 68 40]
 [65 60 40]
 [84 60  0]]
[[0.16666667 0.23529412 0.4       ]
 [0.38461538 0.26666667 0.4       ]
 [0.42857143 0.26666667 0.        ]]
[[6.         4.25       2.5       ]
 [2.6        3.75       2.5       ]
 [2.33333333 3.75              inf]]


/tmp/ipython-input-1937380831.py:5: RuntimeWarning: divide by zero encountered in divide
  print(np.divide(b,a))


ARRAY MANIPULATION

In [ ]:
array=np.random.randint(0,10,(2,3))
print(array)
print(array.shape)

[[1 3 3]
 [6 5 8]]
(2, 3)


In [ ]:
#transpose
trans=np.transpose(array)
print(trans)
print(trans.shape)

[[1 6]
 [3 5]
 [3 8]]
(3, 2)


In [ ]:
array=np.random.randint(0,10,(2,3))
print(array)
print(array.shape)

[[8 0 4]
 [1 0 2]]
(2, 3)


In [ ]:
trans2=array.T
print(trans2)
print(trans2.shape)

[[8 1]
 [0 0]
 [4 2]]
(3, 2)


In [ ]:
#reshaping the array
a=np.random.randint(0,10,(2,3))
print(a)
print(a.shape)

[[0 0 7]
 [7 5 2]]
(2, 3)


In [ ]:
b=a.reshape(3,2)
print(b)
print(b.shape)

[[0 0]
 [7 7]
 [5 2]]
(3, 2)


PANDAS LIBRARY

USEFUL FOR DATA PREPROCESSING & ANALYSIS

PANDAS DATA FRAME

PANDAS DATAFRAME IS 2D TABULAR DATA STRCUTURE WITH LABELLED AXES(ROWS AND COLUMNS)

In [ ]:
#importing the pandas library
import pandas as pd


Creating a Pandas Dataframe

In [ ]:
#importing the boston house price data
from sklearn.datasets import load_boston


ImportError: 
`load_boston` has been removed from scikit-learn since version 1.2.

The Boston housing prices dataset has an ethical problem: as
investigated in [1], the authors of this dataset engineered a
non-invertible variable "B" assuming that racial self-segregation had a
positive impact on house prices [2]. Furthermore the goal of the
research that led to the creation of this dataset was to study the
impact of air quality but it did not give adequate demonstration of the
validity of this assumption.

The scikit-learn maintainers therefore strongly discourage the use of
this dataset unless the purpose of the code is to study and educate
about ethical issues in data science and machine learning.

In this special case, you can fetch the dataset from the original
source::

    import pandas as pd
    import numpy as np

    data_url = "http://lib.stat.cmu.edu/datasets/boston"
    raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
    data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
    target = raw_df.values[1::2, 2]

Alternative datasets include the California housing dataset and the
Ames housing dataset. You can load the datasets as follows::

    from sklearn.datasets import fetch_california_housing
    housing = fetch_california_housing()

for the California housing dataset and::

    from sklearn.datasets import fetch_openml
    housing = fetch_openml(name="house_prices", as_frame=True)

for the Ames housing dataset.

[1] M Carlisle.
"Racist data destruction?"
<https://medium.com/@docintangible/racist-data-destruction-113e3eff54a8>

[2] Harrison Jr, David, and Daniel L. Rubinfeld.
"Hedonic housing prices and the demand for clean air."
Journal of environmental economics and management 5.1 (1978): 81-102.
<https://www.researchgate.net/publication/4974606_Hedonic_housing_prices_and_the_demand_for_clean_air>


In [ ]:
import pandas as pd
import numpy as np

data_url = "http://lib.stat.cmu.edu/datasets/boston"
raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
boston_dataset = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
target = raw_df.values[1::2, 2]

In [ ]:
# prompt: I want to store my data set in variable boston_dataset

type(boston_dataset)

numpy.ndarray

In [ ]:
#pandas dataframe
# Define the feature names manually as the loaded data is a numpy array
feature_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT']
boston_df = pd.DataFrame(boston_dataset, columns=feature_names)

In [ ]:
display(boston_df.head())

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [ ]:
boston_df.shape


(506, 13)

importing the data from a csv file to a dataframe

In [ ]:
diabetes_df=pd.read_csv('/content/diabetes.csv')


In [ ]:
type(diabetes_df)

pandas.core.frame.DataFrame

In [ ]:
diabetes_df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [ ]:
diabetes_df.shape


Loading the data from a excel file to a a pandas dataframe

pd.read_excel('file_path')

In [ ]:
boston_df.to_csv('boston.csv')

Exporting the pandas dataframe to an excel file
df.to_excel('filename')

In [ ]:
#creating a pandas dataframe as random values
random_df=pd.DataFrame(np.random.rand(20,10))  # 20 rows and 10 columns all have values b/w 0 annd 1

In [ ]:
random_df.head()

0         1         2         3         4         5         6  \
0  0.209361  0.512315  0.559334  0.674092  0.702198  0.549237  0.451926   
1  0.739683  0.851825  0.480651  0.110954  0.744514  0.505029  0.454036   
2  0.813741  0.736319  0.028521  0.363085  0.619896  0.150040  0.128841   
3  0.924073  0.607858  0.681028  0.342340  0.458103  0.226550  0.449916   
4  0.920911  0.218448  0.748314  0.541674  0.881405  0.462342  0.834000   

          7         8         9  
0  0.587652  0.681701  0.573559  
1  0.595530  0.315140  0.593048  
2  0.045270  0.552758  0.050130  
3  0.660418  0.861635  0.309933  
4  0.513546  0.999898  0.219426

In [ ]:
random_df.shape

Inspecting DataFrame

In [ ]:
#finding the number of rows and columns
boston_df.shape

(506, 13)

In [ ]:
#print the 1st 5 rows of the dataset
boston_df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [ ]:
#prints the last 5 rows of the data_set
boston_df.tail()

CRIM   ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
501  0.06263  0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527  0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076  0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959  0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741  0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88

In [ ]:
#gives various info about data_set
#here for each column which is 13 it is giving corresponding values in 505 rows which are not null
#means there is no missing values in dataset
boston_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
dtypes: float64(13)
memory usage: 51.5 KB


In [ ]:
#finding the number of missing values
boston_df.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
dtype: int64

In [ ]:
diabetes_df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [ ]:
#counting the values based on labels
diabetes_df.value_counts('Outcome')

Outcome
0    500
1    268
Name: count, dtype: int64

In [ ]:
diabetes_df.value_counts('Age')

Age
22    72
21    63
25    48
24    46
23    38
28    35
26    33
27    32
29    29
31    24
41    22
30    21
37    19
42    18
33    17
32    16
38    16
36    16
45    15
34    14
46    13
43    13
40    13
39    12
35    10
44     8
51     8
50     8
52     8
58     7
54     6
47     6
49     5
53     5
57     5
48     5
60     5
55     4
63     4
62     4
66     4
56     3
65     3
59     3
67     3
61     2
69     2
64     1
68     1
70     1
72     1
81     1
Name: count, dtype: int64

In [ ]:
#group the values based on the mean  #
diabetes_df.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

The person who are not diabetic have glucose level 141.257.. they have average of bloodPressure 70.82 and person who are diabetic have a average age of 37 years

these type of info we can extract

STATISTICAL MEASURE

In [ ]:
#COUNT )R number of values
boston_df.count()

CRIM       506
ZN         506
INDUS      506
CHAS       506
NOX        506
RM         506
AGE        506
DIS        506
RAD        506
TAX        506
PTRATIO    506
B          506
LSTAT      506
dtype: int64

In [ ]:
boston_df.mean()

CRIM         3.613524
ZN          11.363636
INDUS       11.136779
CHAS         0.069170
NOX          0.554695
RM           6.284634
AGE         68.574901
DIS          3.795043
RAD          9.549407
TAX        408.237154
PTRATIO     18.455534
B          356.674032
LSTAT       12.653063
dtype: float64

In [ ]:
#standard deviation
boston_df.std()

CRIM         8.601545
ZN          23.322453
INDUS        6.860353
CHAS         0.253994
NOX          0.115878
RM           0.702617
AGE         28.148861
DIS          2.105710
RAD          8.707259
TAX        168.537116
PTRATIO      2.164946
B           91.294864
LSTAT        7.141062
dtype: float64

In [ ]:
#mini value
boston_df.min()



In [ ]:
#maximum values
boston_df.max()

In [ ]:
boston_df.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT  
count  506.000000  
mean    12.653063  
std      7.141062  
min      1.730000  
25%      6.950000  
50%     11.360000  
75%     16.955000  
max     37.970000

 FOR column CRIM --> it has 506 rows,in all 505 values it has a mean value of 3.6 standard deviation in 506 datapoints of crim column it has standard deviation of 8.6
 it has min value of 0.00632
 it has max value of 88.976
 25% value in 505 values are less than 0.08245
 50% values in 505 values are less than 0.256510
 75% values............
 this type of info we can extract from this method

Manipulating a dataframe


In [ ]:
#adding a column to a dataframe
boston_df['Price']=target

This code creates a new column named 'Price' in the `boston_df` DataFrame and populates it with the values from the `target` array.

In [ ]:
boston_df.head() # this dataset is used to predict the price of house when given following features

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [ ]:
#removing a particular row or column
boston_df.drop(index=0 ,axis=0) # axis =0 when to remove a row and axis =1 when to remove a column
#index=0 means 1st row
#it creates a new dataframe and delete 1st row and store the left rows we have to write inplace=True for permanent deletion
# temporary remove store it in another variable to delete permanently or use any other method

CRIM   ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
1    0.02731  0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729  0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237  0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905  0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
5    0.02985  0.0   2.18   0.0  0.458  6.430  58.7  6.0622  3.0  222.0   
..       ...  ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263  0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527  0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076  0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959  0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741  0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  Price  
1       17.8  396.90   9.14   21.6  
2       17.8  392.83   4.03   34.7  
3       18.7  394.63   2.94   33.4  
4       18.7  396.90   5.33   36.2  
5       18.7  394.12   5.21   28.7  
..       ...     ...    ...    ...  
501     21.0  391.99   9.67   22.4  
502     21.0  396.90   9.08   20.6  
503     21.0  396.90   5.64   23.9  
504     21.0  393.45   6.48   22.0  
505     21.0  396.90   7.88   11.9  

[505 rows x 14 columns]

In [ ]:
#drop a column
boston_df.drop(columns='ZN',axis=1) # temporary remove store it in another variable to delete permanently or use any other method

CRIM  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  PTRATIO  \
0    0.00632   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0     15.3   
1    0.02731   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0     17.8   
2    0.02729   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0     17.8   
3    0.03237   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0     18.7   
4    0.06905   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0     18.7   
..       ...    ...   ...    ...    ...   ...     ...  ...    ...      ...   
501  0.06263  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0     21.0   
502  0.04527  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0     21.0   
503  0.06076  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0     21.0   
504  0.10959  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0     21.0   
505  0.04741  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0     21.0   

          B  LSTAT  Price  
0    396.90   4.98   24.0  
1    396.90   9.14   21.6  
2    392.83   4.03   34.7  
3    394.63   2.94   33.4  
4    396.90   5.33   36.2  
..      ...    ...    ...  
501  391.99   9.67   22.4  
502  396.90   9.08   20.6  
503  396.90   5.64   23.9  
504  393.45   6.48   22.0  
505  396.90   7.88   11.9  

[506 rows x 13 columns]

In [ ]:
#locate particular row or particular column
boston_df.iloc[2]#row

CRIM         0.02729
ZN           0.00000
INDUS        7.07000
CHAS         0.00000
NOX          0.46900
RM           7.18500
AGE         61.10000
DIS          4.96710
RAD          2.00000
TAX        242.00000
PTRATIO     17.80000
B          392.83000
LSTAT        4.03000
Price       34.70000
Name: 2, dtype: float64

In [ ]:
#loacting particular columns
print(boston_df.iloc[:0]) #first column
print(boston_df.iloc[:1]) #second column
print(boston_df.iloc[:2]) #3rd column
print(boston_df.iloc[:-1]) #last columns

 CORRELATION

 1.POSITIVE CORRELATION
 2.NEGATIVE CORRELATION

 13 columns means 13 variable and how these variables are correlated to each other is shown using correlation
 means in boston price predicting dataset
 price and crime are negatively correlated with each other as if crime increases then price decreases. But no of rooms and price are positively correlated as if no. of rooms inc. then price also inc.


In [ ]:
import pandas as pd
import numpy as np

data_url = "http://lib.stat.cmu.edu/datasets/boston"
raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
boston_dataset = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
target = raw_df.values[1::2, 2]

In [ ]:
# pandas dataframe
# Define the feature names manually as the loaded data is a numpy array
feature_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'PRICE']
boston_df = pd.DataFrame(boston_dataset, columns=feature_names[:-1]) # Create DataFrame with original features
boston_df['PRICE'] = target # Add the 'PRICE' column

In [ ]:
boston_df.corr()

CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
CRIM     1.000000 -0.200469  0.406583 -0.055892  0.420972 -0.219247  0.352734   
ZN      -0.200469  1.000000 -0.533828 -0.042697 -0.516604  0.311991 -0.569537   
INDUS    0.406583 -0.533828  1.000000  0.062938  0.763651 -0.391676  0.644779   
CHAS    -0.055892 -0.042697  0.062938  1.000000  0.091203  0.091251  0.086518   
NOX      0.420972 -0.516604  0.763651  0.091203  1.000000 -0.302188  0.731470   
RM      -0.219247  0.311991 -0.391676  0.091251 -0.302188  1.000000 -0.240265   
AGE      0.352734 -0.569537  0.644779  0.086518  0.731470 -0.240265  1.000000   
DIS     -0.379670  0.664408 -0.708027 -0.099176 -0.769230  0.205246 -0.747881   
RAD      0.625505 -0.311948  0.595129 -0.007368  0.611441 -0.209847  0.456022   
TAX      0.582764 -0.314563  0.720760 -0.035587  0.668023 -0.292048  0.506456   
PTRATIO  0.289946 -0.391679  0.383248 -0.121515  0.188933 -0.355501  0.261515   
B       -0.385064  0.175520 -0.356977  0.048788 -0.380051  0.128069 -0.273534   
LSTAT    0.455621 -0.412995  0.603800 -0.053929  0.590879 -0.613808  0.602339   
PRICE   -0.388305  0.360445 -0.483725  0.175260 -0.427321  0.695360 -0.376955   

              DIS       RAD       TAX   PTRATIO         B     LSTAT     PRICE  
CRIM    -0.379670  0.625505  0.582764  0.289946 -0.385064  0.455621 -0.388305  
ZN       0.664408 -0.311948 -0.314563 -0.391679  0.175520 -0.412995  0.360445  
INDUS   -0.708027  0.595129  0.720760  0.383248 -0.356977  0.603800 -0.483725  
CHAS    -0.099176 -0.007368 -0.035587 -0.121515  0.048788 -0.053929  0.175260  
NOX     -0.769230  0.611441  0.668023  0.188933 -0.380051  0.590879 -0.427321  
RM       0.205246 -0.209847 -0.292048 -0.355501  0.128069 -0.613808  0.695360  
AGE     -0.747881  0.456022  0.506456  0.261515 -0.273534  0.602339 -0.376955  
DIS      1.000000 -0.494588 -0.534432 -0.232471  0.291512 -0.496996  0.249929  
RAD     -0.494588  1.000000  0.910228  0.464741 -0.444413  0.488676 -0.381626  
TAX     -0.534432  0.910228  1.000000  0.460853 -0.441808  0.543993 -0.468536  
PTRATIO -0.232471  0.464741  0.460853  1.000000 -0.177383  0.374044 -0.507787  
B        0.291512 -0.444413 -0.441808 -0.177383  1.000000 -0.366087  0.333461  
LSTAT   -0.496996  0.488676  0.543993  0.374044 -0.366087  1.000000 -0.737663  
PRICE    0.249929 -0.381626 -0.468536 -0.507787  0.333461 -0.737663  1.000000